File: src/main/java/by/sorface/sorface/gateway/config/handlers/HttpStatusJsonServerAuthenticationEntryPoint.java
package by.sorface.sorface.gateway.config.handlers;

import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.server.ServerAuthenticationEntryPoint;
import org.springframework.util.Assert;
import org.springframework.web.server.ServerWebExchange;
import reactor.core.publisher.Mono;

import static by.sorface.sorface.gateway.utils.JsonHttpResponseUtils.toJsonException;

public class HttpStatusJsonServerAuthenticationEntryPoint implements ServerAuthenticationEntryPoint {

    private final HttpStatus httpStatus;

    public HttpStatusJsonServerAuthenticationEntryPoint(HttpStatus httpStatus) {
        this.httpStatus = httpStatus;
        Assert.notNull(httpStatus, "httpStatus cannot be null");
    }

    @Override
    public Mono<Void> commence(ServerWebExchange exchange, AuthenticationException ex) {
        return Mono.just(exchange.getResponse()).flatMap(response -> {
            response.setStatusCode(httpStatus);
            response.getHeaders().setContentType(MediaType.APPLICATION_JSON);

            return toJsonException(response, httpStatus, ex);
        });
    }

}

File: src/main/kotlin/by/sorface/gateway/config/entrypoints/HttpStatusJsonServerAuthenticationEntryPoint.kt
package by.sorface.gateway.config.entrypoints

import by.sorface.gateway.utils.JsonHttpResponseUtils
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.server.ServerAuthenticationEntryPoint
import org.springframework.util.Assert
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Mono

class HttpStatusJsonServerAuthenticationEntryPoint(
    private val httpStatus: HttpStatus
) : ServerAuthenticationEntryPoint {

    init {
        Assert.notNull(httpStatus, "HTTP Status cannot be null")
    }

    override fun commence(exchange: ServerWebExchange, ex: AuthenticationException): Mono<Void> {
        return Mono.just(exchange.response)
            .flatMap { response ->
                response.setStatusCode(httpStatus)
                response.headers.contentType = MediaType.APPLICATION_JSON
                return@flatMap JsonHttpResponseUtils.buildJsonAuthorizationError(response, httpStatus, ex)
            }
    }
}
